orchestrator/test_runner/src/ica.rs

```diff
@@ -0,0 +1,816 @@
+pub const STAKING_DELEGATE_TYPE_URL: &str = "/cosmos.staking.v1beta1.MsgDelegate";
+
+// Trait to serialize/deserialize types to and from `prost_types::Any`
+pub trait AnyConvert: Sized {
+    // Deserialize value from `prost_types::Any`
+    fn from_any(value: &Any) -> Result<Self, DecodeError>;
+
+    // Serialize value to `prost_types::Any`
+    fn to_any(&self) -> Result<Any, EncodeError>;
+}
+
+// Encodes a message into protobuf
+pub fn proto_encode<M: Message>(message: &M) -> Result<Vec<u8>, EncodeError> {
+    let mut buf = Vec::with_capacity(message.encoded_len());
+    message
+        .encode(&mut buf)
+        .expect("unable to encode protobuf message");
+    Ok(buf)
+}
+
+impl AnyConvert for MsgDelegate {
+    fn from_any(value: &Any) -> Result<Self, DecodeError> {
+        assert_eq!(
+            value.type_url, STAKING_DELEGATE_TYPE_URL,
+            "invalid type url for `Any` type: expected `{}` and found `{}`",
+            STAKING_DELEGATE_TYPE_URL, value.type_url
+        );
+
+        <Self as ::prost::Message>::decode(value.value.as_slice()).map_err(Into::into)
+    }
+
+    fn to_any(&self) -> Result<Any, EncodeError> {
+        Ok(::prost_types::Any {
+            type_url: STAKING_DELEGATE_TYPE_URL.to_owned(),
+            value: proto_encode(self)?,
+        })
+    }
+}
+
+impl AnyConvert for MsgSendToEth {
+    fn from_any(value: &Any) -> Result<Self, DecodeError> {
+        assert_eq!(
+            value.type_url, MSG_SEND_TO_ETH_TYPE_URL,
+            "invalid type url for `Any` type: expected `{}` and found `{}`",
+            MSG_SEND_TO_ETH_TYPE_URL, value.type_url
+        );
+
+        <Self as ::prost::Message>::decode(value.value.as_slice()).map_err(Into::into)
+    }
+
+    fn to_any(&self) -> Result<Any, EncodeError> {
+        Ok(Any {
+            type_url: MSG_SEND_TO_ETH_TYPE_URL.to_owned(),
+            value: proto_encode(self)?,
+        })
+    }
+}
+
+/// Test Interchain accounts host / controller. Create , Send , Delegate, Send_to_Eth
+/// Plan is
+/// 1. Create and pass proposal to allow all ica_host messages <fn add_ica_host_allow_messages>
+/// 2. Get connection_id , cpc_connection_id:  <fn get_connection_id>
+/// 3. Register interchain account gravity -> ibc: <fn create_interchain_account>
+/// 4. Check account registered: <fn get_interchain_account>
+/// 5. Send some stake tokens: <fn send_tokens_to_interchain_account>, <fn get_interchain_account_balance>
+/// 6. Delegate <fn delegate_from_interchain_account>
+/// 7. Check delegations <fn check_delegatinons>
+/// 8. Adopt footoken metadata
+/// 9. SubmitTx MsgSendToEth from counterpartychain ICA. <fn submit_tx>
+/// 10. Request & submit batch
+/// 11. Check eth balance
+pub async fn ica_test(
+    contact: &Contact,
+    keys: Vec<ValidatorKeys>,
+    ibc_keys: Vec<CosmosPrivateKey>,
+    gravity_address: EthAddress,
+    web30: &Web3,
+    grpc_client: GravityQueryClient<Channel>,
+) {
```

Comment:
This test needs just a little bit of work to clarify what is happening when viewing the logs, but it is a great test of base ICA functionality.

To summarize the test, it initializes the gravity and ibc chain's ICA Host AllowMessages parameter, creates an interchain account on each of gravity and the ibc chain, sends those interchain accounts some "stake" funds with a basic bank/MsgSend, and uses ICA to delegate to validator 0 from gravity and from ibc-chain, respectively.

Reply:
Also send to eth from ibc chain ( if MsgSendToEth added to codec on ibc chain  )